### ISSUE TITLE
cat: /dev/fd/63: No such file or directory

### ISSUE THREAD

**Issue by HoracyAmadeusz** [rvm/rvm#3950]
## Steps to reproduce

$ mkdir tmp
$ cd tmp
$ gpg --keyserver hkp://keys.gnupg.net --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3
$ curl -sSL https://get.rvm.io | bash -s stable
$ source ~/.rvm/scripts/rvm
$ cd ~
/home/USER/tmp
cat: /dev/fd/63: No such file or directory

## Environment info

FreeBSD 10.3-STABLE
GNU bash, version 4.4.5(0)-release (amd64-portbld-freebsd10.3)

**Comment by avkarenow** [comment]
I have the same problem - every run of 'cd' command shows this error: cat: /dev/fd/63: No such file or directory. I tried it on fresh FreeBSD 10.3 installation with newest rvm stable version as normal system user (not root).

Any solutions?

**Comment by mpapis** [comment]
What `bash --version` are you using?

**Comment by avkarenow** [comment]
$ bash --version
GNU bash, version 4.4.5(0)-release (amd64-portbld-freebsd10.3)

**Comment by HoracyAmadeusz** [comment]
$ bash --version
GNU bash, wersja 4.4.5(0)-release (amd64-portbld-freebsd10.3)

**Comment by adeadman** [comment]
I found [this comment](https://github.com/andsens/homeshick/issues/124#issuecomment-63984418) which suggests that Bash now needs fdescfs mounted at `/dev/fd` to function properly, but it is not done by FreeBSD by default.

**Comment by mpapis** [comment]
@adeadman I wonder if this can be checked so we can add a warning, my Ubuntu uses older Bash so I do not know yet how to check it - can you try `mount | grep "on /dev"`?

**Comment by HoracyAmadeusz** [comment]
@mpapis 
yes, it's not mounted in FreeBSD-10.3 by default bash installation.

**Comment by HoracyAmadeusz** [comment]
@mpapis 
As far as I know mounting fdescfs is a security risc and my shell hosting provider is not going to mount it on his servers because of that. I think it may be an issue in case of many other hosting providers and the reason why it is not mounted by default in case of FreeBSD operating system.

**Comment by HoracyAmadeusz** [comment]
@mpapis 
Can current script be adapted to work without fdescfs to fix that?

**Comment by Oxylibrium** [comment]
The cPanel I use also doesn't mount fdescfs. Fix?

**Comment by HoracyAmadeusz** [comment]
Any hope to get this fixed?

**Comment by mpapis** [comment]
this will require bigger investigation on our side (even with replicating the conditions) - and then rewriting some of the functionalities, this will take time, lots of time -> @pkuczynski WDYT?

**Comment by avkarenow** [comment]
Any news?

**Comment by pkuczynski** [comment]
@mpapis what would be involved in this? If this is a bash issue then I can’t imagine us rewriting from bash to something else...

The simplest thing would be to display a warning...

**Comment by zhanhb** [comment]
```shell
ln -s /proc/self/fd /dev/fd
```

**Comment by avkarenow** [comment]
@zhanhb it is not a solution.

On the system where I have an account there is not /proc mounted:
> $ ls /proc/
> $

**Comment by sriccio** [comment]
I had the very same issue in a jailed environnement without access to `/proc`, however I nailed down the issue to the file `/usr/local/rvm/scripts/hook`.

```
    __rvm_read_lines _hooks_list <(
      __rvm_find -L "${hooks[@]}" -iname "$rvm_hook*" -type f 2>/dev/null || true
    )
```

This part of the code use command substitution, if I'm correct about it, the `<( .... )` triggers it.

Commenting out this part and the warning/error is gone, but it also then will not load any hooks anymore.
Maybe is there an alternative way to write this part without using command substitution ?

**Comment by xiangdeyue** [comment]
> I had the very same issue in a jailed environnement without access to `/proc`, however I nailed down the issue to the file `/usr/local/rvm/scripts/hook`.
> 
> ```
>     __rvm_read_lines _hooks_list <(
>       __rvm_find -L "${hooks[@]}" -iname "$rvm_hook*" -type f 2>/dev/null || true
>     )
> ```
> 
> This part of the code use command substitution, if I'm correct about it, the `<( .... )` triggers it.
> 
> Commenting out this part and the warning/error is gone, but it also then will not load any hooks anymore. Maybe is there an alternative way to write this part without using command substitution ?

    __rvm_read_lines _hooks_list <(
      __rvm_find -L "${hooks[@]}" -iname "$rvm_hook*" -type f 2>/dev/null || true
    ) 2>/dev/null || true
